libsweep/src/sweep.cc

```diff
@@ -100,33 +109,60 @@ void sweep_device_start_scanning(sweep_device_s device, sweep_error_s* error) {
   if (device->is_scanning)
     return;
 
-  sweep::protocol::error_s protocolerror = nullptr;
-  sweep::protocol::write_command(device->serial, sweep::protocol::DATA_ACQUISITION_START, &protocolerror);
-
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to send start scanning command");
-    sweep::protocol::error_destruct(protocolerror);
+  // Get the current motor speed setting
+  sweep_error_s speederror = nullptr;
+  int32_t speed = sweep_device_get_motor_speed(device, &speederror);
+  if (speederror) {
+    *error = sweep_error_construct("unable to start scanning. could not verify motor speed.");
+    sweep_error_destruct(speederror);
     return;
   }
+  // Check that the motor is not stationary
+  if (speed == 0 /*Hz*/) {
+    // If the motor is stationary, adjust it to default 5Hz
+    sweep_device_set_motor_speed(device, 5 /*Hz*/, &speederror);
+    if (speederror) {
+      *error = sweep_error_construct("unable to start scanning. failed to start motor.");
+      sweep_error_destruct(speederror);
+      return;
+    }
+  }
 
-  sweep::protocol::response_header_s response;
-  sweep::protocol::read_response_header(device->serial, sweep::protocol::DATA_ACQUISITION_START, &response, &protocolerror);
+  // Make sure the motor is stabilized so the DS command doesn't fail
+  sweep_error_s stabilityerror = nullptr;
+  sweep_device_wait_until_motor_ready(device, &stabilityerror);
+  if (stabilityerror) {
+    *error = sweep_error_construct("unable to start scanning. motor stability could not be verified.");
+    sweep_error_destruct(stabilityerror);
+    return;
+  }
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to receive start scanning command response");
-    sweep::protocol::error_destruct(protocolerror);
+  // Attempt to start scanning
+  sweep_error_s starterror = nullptr;
+  sweep_device_attempt_start_scanning(device, &starterror);
+  if (starterror) {
+    *error = sweep_error_construct("unable to start scanning.");
+    sweep_error_destruct(starterror);
     return;
   }
 
+  // Start SCAN WORKER
+  device->scan_queue->flush();
   device->is_scanning = true;
+  // START background worker thread
+  device->stop_thread = false;
+  // create a thread
+  std::thread th = std::thread(sweep_device_accumulate_scans, device);
+  // detach the thread so that it runs in the background and cleans itself up
+  th.detach();
```

Comment:
As I understand it, the detached background thread cleans itself up properly when it's method returns or terminates. Seeing how the behavior of the thread exists in the "background" with no need to synchronize, it seems like a much simpler solution to avoid explicitly managing the thread.